### PR TITLE
Split bucket aggregation KQL into sub queries

### PR DESCRIPTION
**Pull request by dupuyjs**
The following changes are proposed:

Preparation work to implement TopHits Aggregation (and nested aggregation).
All bucket aggregation queries are now split into two sub queries (_extdata and _summarizablemetrics).

DateHistogramAggregation Aggregation
```
let _extdata = _data
| extend ['10'] = startofmonth(['timestamp']);
let _summarizablemetrics = _extdata
| summarize count() by ['10']
| order by ['10'] asc;
```

Terms Aggregation
```
let _extdata = _data
| extend ['10'] = ['Carrier'];
let _summarizablemetrics = _extdata
| summarize ['2']=avg(['DistanceMiles']),['11']=max(['dayOfWeek']),count() by ['10'] = ['Carrier']
| order by ['count_'] desc\n| limit 5;"
```

Range and DateRange Aggregations
```
let _extdata = _data
| extend ['2'] = pack_array("range1", "range2", "range3"), ['_range_value'] = pack_array(expr1, expr2, expr3)
| mv-expand ['2'] to typeof(string), ['_range_value']
| where ['_range_value'] == true;
let _summarizablemetrics = _extdata
| summarize count() by ['2']
| order by ['2'] asc;"
```

Filters Aggregation
```
let _extdata = _data
| extend ['2'] = pack_array('k1'), ['_filter_value'] = pack_array(expr1)
| mv-expand ['2'] to typeof(string), ['_filter_value']
| where ['_filter_value'] == true;
let _summarizablemetrics = _extdata
| summarize count() by ['2']
| order by ['2'] asc;"
```

In order to unify the logic even for standalone aggregates, a DefaultAggregation has been created and returns:

```
let _extdata = _data
| extend ['Guid'] = True;
let _summarizablemetrics = _extdata
| summarize count() by ['Guid'];
```

### CODE REVIEW

File: K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs
Comment (eladiw) [inline K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs:31]: same comment for range expression?

File: K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs
Comment (eladiw) [inline K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs:4]: This file was completely changed... i guess since the union part didn't work well with the new approach..

**Review by eladiw — APPROVED** [review]
Approved with comments. please address / comment.
Lets sync BEFORE merge

File: K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs
Comment (AsafMah) [inline K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs:56]: Now DateRange, Range and Filter all have very similar code. Can we generalize it and avoid repetition?

File: K2Bridge.Tests.UnitTests/Visitors/DateHistogramAggregationVisitorTests.cs
Comment (AsafMah) [inline K2Bridge.Tests.UnitTests/Visitors/DateHistogramAggregationVisitorTests.cs:54]: Do we need all of the newlines in the query? They're not meant to be readable anyway

File: K2Bridge/KustoDAL/KustoResponseParser.cs
Comment (dupuyjs) [reply K2Bridge/KustoDAL/KustoResponseParser.cs:179]: Yes. In some situation, table returns columns but no rows (for instance for standalone aggregation out of ranges). AddAggregates expect to read columns name if the value is empty or NaN (for instance for percentiles). But, if rows are empty, we don't enter in AddAggregates. I currently fixed this issue by adding an Empty Row, but it perhaps more appropriate to pass the Table to AddAggregates and rework the logic. I suggest to keep this current fix and open a parallel discussion / workItem to check if there is a more appropriate way to address this point.

File: K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs:37]: It's not an issue in the current context. But when I will add partition, this guid will be used as partition key (and to join tables).

```
let _data = database("dev").kibana_data_flights;
let _extdata = _data
| extend ['eaea70c1-3410-40f1-9400-af2b41d41be0'] = true;
let _summarizablemetrics = _extdata
| summarize ['2']=avg(['DistanceMiles']),count() by ['eaea70c1-3410-40f1-9400-af2b41d41be0'];
let _tophits2 = _extdata
| join kind=inner _summarizablemetrics on ['eaea70c1-3410-40f1-9400-af2b41d41be0']
| partition by ['eaea70c1-3410-40f1-9400-af2b41d41be0']
(
    top 1 by timestamp desc
    | project ['count_'],['3']=pack('DistanceMiles', ['DistanceMiles'], 'timestamp', ['timestamp'])
    | summarize take_any(['count_']),['3%tophits']=make_list(['3']) );
(_tophits2 | as aggs);
```

Comment (eladiw) [reply K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs:37]: so this will be the second phase?

**Comment by eladiw** [comment]
It looks good. After you resolve @AsafMah  commentss and get approval you can merge

File: K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs:31]: Yes. RangeAggregationVisitor and DateRangeAggregationVisitor are similar. Difference is between RangeAggregationExpressionVisitor and DateRangeAggregationExpressionVisitor.

File: K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs:4]: Yes. @tomconte updated DateRange and Filters, and I updated Range to use the same pattern.

File: K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs:37]: Yes, next PR will provide the partition implementation.

File: K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs:56]: Yes, I agree. Especially code between Range and DateRange. But the associated Expression (RangeAggregationVisitor and DateRangeAggregationVisitor) are different. Not so evident to share this code. Do you have a suggestion ?

File: K2Bridge.Tests.UnitTests/Visitors/DateHistogramAggregationVisitorTests.cs
Comment (dupuyjs) [reply K2Bridge.Tests.UnitTests/Visitors/DateHistogramAggregationVisitorTests.cs:54]: The 'newline' logic was really different from each individual Visitor. Some were using only simple command separator |, some were using KustoOperator.CommandSeparator (\n|) which includes newline. And I introduced KustoOperator.Newline (to be used only before let commands).

We can decide to remove all the \n if you agree. I don't have a real preference. The only situation where it helps, is copy/paste from the debugger. If I have your Go, I remove them.

File: K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs
Comment (AsafMah) [reply K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs:56]: Have BaseRangeAggregation that will have the `Field` and `Ranges` props, and move all of the logic to a function that will accept it? 

And if you want to go a step further make a base class that also includes Filter, with virtual members and methods to account for the small differences.

I really don't think the code should be duplicated 3 times, they are very very similar.

File: K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs
Comment (eladiw) [reply K2Bridge/Visitors/Aggregations/RangeAggregationVisitor.cs:31]: I meant, can you add a similar comment like you added above to the extendExpression.

// Range expression:
            // >> abc...

File: K2Bridge/KustoDAL/KustoResponseParser.cs
Comment (AsafMah) [reply K2Bridge/KustoDAL/KustoResponseParser.cs:179]: I guess my question is, if we actually got a table that has just one empty row, will it break/change something?

I don't think it would, but I want to be sure.

Comment (dupuyjs) [reply K2Bridge/KustoDAL/KustoResponseParser.cs:179]: No, I ensure it does not break anything. I tested this hack and it offers only benefits.

File: K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs:56]: OK, l work on this point.

File: K2Bridge/Visitors/Aggregations/DefaultAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/DefaultAggregationVisitor.cs]: Fixed.

File: K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/DateRangeAggregationVisitor.cs:56]: @AsafMah If possible, I prefer to postpone this refactoring until @tomconte implemented changes on filters. The same fix should be applied on Range and DateRange. We will probably have the same fix on all aggregations, but I prefer to be sure it will be the case before starting this code change.

File: K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs
Comment (AsafMah) [inline K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs:54]:
```suggestion
query.Append(projectAwayExpression)
```

File: K2Bridge/Visitors/Aggregations/TermsAggregationVisitor.cs
Comment (AsafMah) [inline K2Bridge/Visitors/Aggregations/TermsAggregationVisitor.cs:33]: We don't really need to encode fields we know ahead of time

**Review by AsafMah — APPROVED** [review]
Looks good!

Minor minor comments.

And also we need to document/write somewhere to unify the logic of Filters and Range when both PRs are merged

File: K2Bridge/Visitors/Aggregations/TermsAggregationVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/TermsAggregationVisitor.cs:33]: Yes, I agree this is not really needed. But I just considered it offers clarity in the KQL produced, everything is encoded properly.

File: K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs
Comment (dupuyjs) [reply K2Bridge/Visitors/Aggregations/AggregationDictionaryVisitor.cs]: Fixed.

**Pull request merged**
